### PR TITLE
New App: Json Display

### PR DESCRIPTION
**Pull request by thickey256**
# Description
A new app for displaying simple json content from a server

# Copilot

### Summary
📄🌟📝

This pull request adds a new app called `jsondisplay`, which can display any json data on the Tidbyt device. It includes the app code, configuration schema, metadata, documentation, and an example json file. This app allows users to customize and visualize their own data sources on the Tidbyt.

> _`json` display app_
> _Starlark code and manifest_
> _README for fall_

### Walkthrough
* Create a new app called json display that can show json data on the Tidbyt device ([link](https://github.com/tidbyt/community/pull/1517/files?diff=unified&w=0#diff-1ad4b219b72fbe9c9ebe67f91204a3b6fd8350dbaddc4ad444372f436d3c0aa9L1-R43), [link](https://github.com/tidbyt/community/pull/1517/files?diff=unified&w=0#diff-ea104f048db6c625c400a94b807b8824e25b2b93990f7c12cfa022cb5b0377f2L1-R20), [link](https://github.com/tidbyt/community/pull/1517/files?diff=unified&w=0#diff-b5b37dc3ef5b283a4f317501c26288be2311e026430e33c08eb142f6b384cfcdR1-R71), [link](https://github.com/tidbyt/community/pull/1517/files?diff=unified&w=0#diff-075fdbb0719aacdf45891b34b4077423e3c6acb8a50ba99235973fb03c09a4f4R1-R8))
* Write the main function in `json_display.star` that takes a config object and returns a render.Root object with the app layout and logic ([link](https://github.com/tidbyt/community/pull/1517/files?diff=unified&w=0#diff-b5b37dc3ef5b283a4f317501c26288be2311e026430e33c08eb142f6b384cfcdR1-R71))
* Define the app configuration fields in the get_schema function in `json_display.star`, using schema.Schema and schema.String ([link](https://github.com/tidbyt/community/pull/1517/files?diff=unified&w=0#diff-b5b37dc3ef5b283a4f317501c26288be2311e026430e33c08eb142f6b384cfcdR1-R71))
* Add an `example.json` file with sample json data that the app can display, and use it as a default option for the config object ([link](https://github.com/tidbyt/community/pull/1517/files?diff=unified&w=0#diff-ea104f048db6c625c400a94b807b8824e25b2b93990f7c12cfa022cb5b0377f2L1-R20))
* Add a `manifest.yaml` file with the app metadata and information, such as name, description, author, and icon ([link](https://github.com/tidbyt/community/pull/1517/files?diff=unified&w=0#diff-075fdbb0719aacdf45891b34b4077423e3c6acb8a50ba99235973fb03c09a4f4R1-R8))
* Add a `README.MD` file with documentation and instructions for the app, such as features, limitations, feedback, and example ([link](https://github.com/tidbyt/community/pull/1517/files?diff=unified&w=0#diff-1ad4b219b72fbe9c9ebe67f91204a3b6fd8350dbaddc4ad444372f436d3c0aa9L1-R43))

### CODE REVIEW

**Comment by thickey256** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by thickey256** [comment]
I have just spent a couple of hours adding basic caching to this, but I don't think this is going to work. I was thinking if I used the url as a cached value then the cache would be unique for that feed. But the more I understand the caching the more I don't think it would work.

So in the unlikely event that this does actually get merged, I think the options are maybe as follows :- 

- No caching (which would result in a new call to the URL every time the app is loaded in the rotation?)
- prepend the cached json string variable with an md5 of the URL eg cef1c2581f56eae9531e134908e6b1d3_json_contents_cached. this is ugly but could work
- do something with `pixlet push --installation-id` and run a cron locally which is the 2nd worst option.

I think probably the MD5 is best, as it will stop different installs clashing with each other (unless they use the same URL) and isolate the data from different URLs a little.

It would be cool to get someone else's point of view on this. if it has no chance of getting merged no matter how the caching is executed then I guess there isn't much point. But it would be good to know.

**Comment by matslina** [comment]
Hey. I think this will work just fine with a couple of edits.

The easiest path forward is to start using the built in caching of the http module. You can read more about it [over here](https://discuss.tidbyt.com/t/http-caching-client/5129). Basically, you can completely stop using `cache.star` and instead control caching by passing `ttl_seconds` to `http.get()`.

Using the URL as cache key seems to be the way to go, and that's what the http cache will do for you.

**Comment by noahpodgurski** [comment]
> Hey. I think this will work just fine with a couple of edits.
> 
> The easiest path forward is to start using the built in caching of the http module. You can read more about it [over here](https://discuss.tidbyt.com/t/http-caching-client/5129). Basically, you can completely stop using `cache.star` and instead control caching by passing `ttl_seconds` to `http.get()`.
> 
> Using the URL as cache key seems to be the way to go, and that's what the http cache will do for you.

I'm sure you're tired of saying this lol, I think many people (myself included) use the cache library because it's still listed in the tidbyt docs

**Comment by thickey256** [comment]
Yes, I didn't even think to look elsewhere because it's front and center in the docs. I will make some code changes!

**Comment by matslina** [comment]
Docs have been updated! LMK (or cut a PR) if you spot any remaining docs pushing people to use cache.star.

**Review by matslina — DISMISSED** [review]

File: apps/jsondisplay/json_display.star [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    json_contents = json.decode(rep.body())
+
+    if rep.status_code != 200:
+        fail("Json URL didn't load %d", rep.status_code)
+
+    #Set the font
+    font = "tom-thumb"
+
+    #Sort out the icon (10x10 png works well)
+    icon_image = http.get(json_contents["title_image"])
```
[the remark this answers is not in the document]
Comment (thickey256) [reply apps/jsondisplay/json_display.star]: Great Idea, no clue why I missed that.

**Pull request merged**